### ISSUE TITLE
Extract header options into partials

### ISSUE THREAD

**Issue by sisp** [squidfunk/mkdocs-material#5858]
### Context

I'm working on a customization that requires more than pure stylistic changes. In my case, the header options (color palette and language selector) and search form need to be moved to from the header nav to the tabs nav. I'm trying to achieve this with the least amount of code duplication as any copied code from upstream means technical debt.

### Description

I suggest to extract the header options ([color palette](https://github.com/squidfunk/mkdocs-material/blob/9.1.21/src/partials/header.html#L74-L110) and [language selector](https://github.com/squidfunk/mkdocs-material/blob/9.1.21/src/partials/header.html#L112-L140)) into partials, so I can include them (in my case) in my `tabs.html` override. To hide those components in the header nav, I could either apply `display: none` to their containers, or `mkdocs-material` could additionally wrap them in a Jinja [block which I can override](https://squidfunk.github.io/mkdocs-material/customization/#overriding-blocks) with empty content.

### Related links

- [Requesting a change](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/)
- [Extending the theme](https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme)

### Use Cases

This change would improve structural theme customization beyond style changes that require rendering the mentioned components in a different place.

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by squidfunk** [comment]
Thanks for suggesting. I understand, that it would make customization simpler. However, we're currently not considering doing that, because the more partials you have, the slower the build will become – that's just how Jinja works. It's essentially a trade-off between performance and customization, and the header is used in every site. We always need to keep on eye on build performance, and this would make it slower for everyone.

**Comment by sisp** [comment]
I'm a bit surprised by your anticipated performance degradation because Jinja compiles templates to Python code and caches it.

I've done a little benchmark comparison of inline components vs. included partials for 100 and 1000 pages of documentation, and I don't see a big difference. Wouldn't you say this minor (if at all, there's some variance) build time increase is worth the much better customization, @squidfunk?

## 100 pages @ 50 runs

* Inline components:

    ```text
                Mean        Std.Dev.    Min         Median      Max
    real        1.023       0.065       0.924       1.027       1.172       
    user        0.951       0.065       0.851       0.952       1.118       
    sys         0.077       0.015       0.051       0.074       0.116
    ```

* Included component partials:

    ```text
                Mean        Std.Dev.    Min         Median      Max
    real        1.018       0.064       0.926       1.011       1.193       
    user        0.951       0.065       0.852       0.949       1.110       
    sys         0.073       0.014       0.043       0.073       0.108
    ```

## 1000 pages @ 20 runs

* Inline components:

    ```text
                Mean        Std.Dev.    Min         Median      Max
    real        17.150      0.833       15.716      16.894      19.454      
    user        16.652      0.796       15.347      16.416      18.881      
    sys         0.476       0.046       0.359       0.481       0.541
    ```

* Included component partials:

    ```text
                Mean        Std.Dev.    Min         Median      Max
    real        17.760      1.106       14.886      18.004      19.253      
    user        17.245      1.113       14.463      17.534      18.672      
    sys         0.462       0.048       0.364       0.470       0.542
    ```

---

## Reproduction

<details>
<summary>Script to generate the test docs: <code>gen_docs.py</code></summary>

```python
import json
import sys
from pathlib import Path
from shutil import rmtree
from textwrap import dedent


def _reset():
    rmtree("docs", ignore_errors=True)
    Path("docs").mkdir()


def _gen_page(filename, title):
    Path("docs", filename).write_text(f"# {title}")
    return {title: filename}


def _gen_docs(num_pages):
    nav_entries = [
        _gen_page(f"page-{i}.md", f"Page {i}") for i in range(1, 1 + num_pages)
    ]
    Path("mkdocs.yml").write_text(
        dedent(
            f"""\
            site_name: Test Site
            theme:
              name: material
              palette: 
                - scheme: default
                  toggle:
                    icon: material/brightness-7 
                    name: Switch to dark mode
                - scheme: slate
                  toggle:
                    icon: material/brightness-4
                    name: Switch to light mode
            extra:
              alternate:
                - name: English
                  link: /en/ 
                  lang: en
                - name: Deutsch
                  link: /de/
                  lang: de
            nav:
              - Home: {json.dumps(nav_entries)}
            """
        )
    )


if __name__ == "__main__":
    _reset()
    _gen_docs(int(sys.argv[1]))
```

</details>

<details>
<summary>Patch to apply to <a href="https://github.com/squidfunk/mkdocs-material/commit/a0921bcaf42251ecf92cd8a901c71030229b6c99"><code>a0921bc</code></a>: <code>extract-partials.diff</code></summary>

```diff
diff --git a/material/partials/alternate.html b/material/partials/alternate.html
new file mode 100644
index 000000000..4492d6576
--- /dev/null
+++ b/material/partials/alternate.html
@@ -0,0 +1,22 @@
+{#-
+  This file was automatically generated - do not edit
+-#}
+<div class="md-header__option">
+  <div class="md-select">
+    {% set icon = config.theme.icon.alternate or "material/translate" %}
+    <button class="md-header__button md-icon" aria-label="{{ lang.t('select.language') }}">
+      {% include ".icons/" ~ icon ~ ".svg" %}
+    </button>
+    <div class="md-select__inner">
+      <ul class="md-select__list">
+        {% for alt in config.extra.alternate %}
+          <li class="md-select__item">
+            <a href="{{ alt.link | url }}" hreflang="{{ alt.lang }}" class="md-select__link">
+              {{ alt.name }}
+            </a>
+          </li>
+        {% endfor %}
+      </ul>
+    </div>
+  </div>
+</div>
diff --git a/material/partials/header.html b/material/partials/header.html
index 3b3c52f71..2e1490d64 100644
--- a/material/partials/header.html
+++ b/material/partials/header.html
@@ -35,41 +35,11 @@
     </div>
     {% if config.theme.palette %}
       {% if not config.theme.palette is mapping %}
-        <form class="md-header__option" data-md-component="palette">
-          {% for option in config.theme.palette %}
-            {% set scheme  = option.scheme  | d("default", true) %}
-            {% set primary = option.primary | d("indigo", true) %}
-            {% set accent  = option.accent  | d("indigo", true) %}
-            <input class="md-option" data-md-color-media="{{ option.media }}" data-md-color-scheme="{{ scheme | replace(' ', '-') }}" data-md-color-primary="{{ primary | replace(' ', '-') }}" data-md-color-accent="{{ accent | replace(' ', '-') }}" {% if option.toggle %} aria-label="{{ option.toggle.name }}" {% else %} aria-hidden="true" {% endif %} type="radio" name="__palette" id="__palette_{{ loop.index }}">
-            {% if option.toggle %}
-              <label class="md-header__button md-icon" title="{{ option.toggle.name }}" for="__palette_{{ loop.index0 or loop.length }}" hidden>
-                {% include ".icons/" ~ option.toggle.icon ~ ".svg" %}
-              </label>
-            {% endif %}
-          {% endfor %}
-        </form>
+        {% include "partials/palette.html" %}
       {% endif %}
     {% endif %}
     {% if config.extra.alternate %}
-      <div class="md-header__option">
-        <div class="md-select">
-          {% set icon = config.theme.icon.alternate or "material/translate" %}
-          <button class="md-header__button md-icon" aria-label="{{ lang.t('select.language') }}">
-            {% include ".icons/" ~ icon ~ ".svg" %}
-          </button>
-          <div class="md-select__inner">
-            <ul class="md-select__list">
-              {% for alt in config.extra.alternate %}
-                <li class="md-select__item">
-                  <a href="{{ alt.link | url }}" hreflang="{{ alt.lang }}" class="md-select__link">
-                    {{ alt.name }}
-                  </a>
-                </li>
-              {% endfor %}
-            </ul>
-          </div>
-        </div>
-      </div>
+      {% include "partials/alternate.html" %}
     {% endif %}
     {% if "material/search" in config.plugins %}
       <label class="md-header__button md-icon" for="__search">
diff --git a/material/partials/palette.html b/material/partials/palette.html
new file mode 100644
index 000000000..11ec11158
--- /dev/null
+++ b/material/partials/palette.html
@@ -0,0 +1,16 @@
+{#-
+  This file was automatically generated - do not edit
+-#}
+<form class="md-header__option" data-md-component="palette">
+  {% for option in config.theme.palette %}
+    {% set scheme  = option.scheme  | d("default", true) %}
+    {% set primary = option.primary | d("indigo", true) %}
+    {% set accent  = option.accent  | d("indigo", true) %}
+    <input class="md-option" data-md-color-media="{{ option.media }}" data-md-color-scheme="{{ scheme | replace(' ', '-') }}" data-md-color-primary="{{ primary | replace(' ', '-') }}" data-md-color-accent="{{ accent | replace(' ', '-') }}" {% if option.toggle %} aria-label="{{ option.toggle.name }}" {% else %} aria-hidden="true" {% endif %} type="radio" name="__palette" id="__palette_{{ loop.index }}">
+    {% if option.toggle %}
+      <label class="md-header__button md-icon" title="{{ option.toggle.name }}" for="__palette_{{ loop.index0 or loop.length }}" hidden>
+        {% include ".icons/" ~ option.toggle.icon ~ ".svg" %}
+      </label>
+    {% endif %}
+  {% endfor %}
+</form>
diff --git a/src/partials/alternate.html b/src/partials/alternate.html
new file mode 100644
index 000000000..b84eb4ebf
--- /dev/null
+++ b/src/partials/alternate.html
@@ -0,0 +1,26 @@
+<div class="md-header__option">
+  <div class="md-select">
+    {% set icon = config.theme.icon.alternate or "material/translate" %}
+    <button
+      class="md-header__button md-icon"
+      aria-label="{{ lang.t('select.language') }}"
+    >
+      {% include ".icons/" ~ icon ~ ".svg" %}
+    </button>
+    <div class="md-select__inner">
+      <ul class="md-select__list">
+        {% for alt in config.extra.alternate %}
+          <li class="md-select__item">
+            <a
+              href="{{ alt.link | url }}"
+              hreflang="{{ alt.lang }}"
+              class="md-select__link"
+            >
+              {{ alt.name }}
+            </a>
+          </li>
+        {% endfor %}
+      </ul>
+    </div>
+  </div>
+</div>
diff --git a/src/partials/header.html b/src/partials/header.html
index 59e4cd5b0..4d0502c26 100644
--- a/src/partials/header.html
+++ b/src/partials/header.html
@@ -74,69 +74,13 @@
     <!-- Color palette -->
     {% if config.theme.palette %}
       {% if not config.theme.palette is mapping %}
-        <form class="md-header__option" data-md-component="palette">
-          {% for option in config.theme.palette %}
-            {% set scheme  = option.scheme  | d("default", true) %}
-            {% set primary = option.primary | d("indigo", true) %}
-            {% set accent  = option.accent  | d("indigo", true) %}
-            <input
-              class="md-option"
-              data-md-color-media="{{ option.media }}"
-              data-md-color-scheme="{{ scheme | replace(' ', '-') }}"
-              data-md-color-primary="{{ primary | replace(' ', '-') }}"
-              data-md-color-accent="{{ accent | replace(' ', '-') }}"
-              {% if option.toggle %}
-                aria-label="{{ option.toggle.name }}"
-              {% else  %}
-                aria-hidden="true"
-              {% endif %}
-              type="radio"
-              name="__palette"
-              id="__palette_{{ loop.index }}"
-            />
-            {% if option.toggle %}
-              <label
-                class="md-header__button md-icon"
-                title="{{ option.toggle.name }}"
-                for="__palette_{{ loop.index0 or loop.length }}"
-                hidden
-              >
-                {% include ".icons/" ~ option.toggle.icon ~ ".svg" %}
-              </label>
-            {% endif %}
-          {% endfor %}
-        </form>
+        {% include "partials/palette.html" %}
       {% endif %}
     {% endif %}
 
     <!-- Site language selector -->
     {% if config.extra.alternate %}
-      <div class="md-header__option">
-        <div class="md-select">
-          {% set icon = config.theme.icon.alternate or "material/translate" %}
-          <button
-            class="md-header__button md-icon"
-            aria-label="{{ lang.t('select.language') }}"
-          >
-            {% include ".icons/" ~ icon ~ ".svg" %}
-          </button>
-          <div class="md-select__inner">
-            <ul class="md-select__list">
-              {% for alt in config.extra.alternate %}
-                <li class="md-select__item">
-                  <a
-                    href="{{ alt.link | url }}"
-                    hreflang="{{ alt.lang }}"
-                    class="md-select__link"
-                  >
-                    {{ alt.name }}
-                  </a>
-                </li>
-              {% endfor %}
-            </ul>
-          </div>
-        </div>
-      </div>
+      {% include "partials/alternate.html" %}
     {% endif %}
 
     <!-- Button to open search modal -->
diff --git a/src/partials/palette.html b/src/partials/palette.html
new file mode 100644
index 000000000..fe6a1363c
--- /dev/null
+++ b/src/partials/palette.html
@@ -0,0 +1,32 @@
+<form class="md-header__option" data-md-component="palette">
+  {% for option in config.theme.palette %}
+    {% set scheme  = option.scheme  | d("default", true) %}
+    {% set primary = option.primary | d("indigo", true) %}
+    {% set accent  = option.accent  | d("indigo", true) %}
+    <input
+      class="md-option"
+      data-md-color-media="{{ option.media }}"
+      data-md-color-scheme="{{ scheme | replace(' ', '-') }}"
+      data-md-color-primary="{{ primary | replace(' ', '-') }}"
+      data-md-color-accent="{{ accent | replace(' ', '-') }}"
+      {% if option.toggle %}
+        aria-label="{{ option.toggle.name }}"
+      {% else  %}
+        aria-hidden="true"
+      {% endif %}
+      type="radio"
+      name="__palette"
+      id="__palette_{{ loop.index }}"
+    />
+    {% if option.toggle %}
+      <label
+        class="md-header__button md-icon"
+        title="{{ option.toggle.name }}"
+        for="__palette_{{ loop.index0 or loop.length }}"
+        hidden
+      >
+        {% include ".icons/" ~ option.toggle.icon ~ ".svg" %}
+      </label>
+    {% endif %}
+  {% endfor %}
+</form>
```

</details>

### Steps

1. Clone `mkdocs-material` and switch to a0921bcaf42251ecf92cd8a901c71030229b6c99:

    ```sh
    git clone git@github.com:squidfunk/mkdocs-material.git /tmp/mkdocs-material
    cd /tmp/mkdocs-material
    git checkout a0921bcaf42251ecf92cd8a901c71030229b6c99
    ```

1. Create a test docs project:

    ```sh
    mkdir /tmp/mkdocs-material-bench-partials
    cd /tmp/mkdocs-material-bench-partials
    ```

1. Install `mkdocs-material` in editable mode:

    ```sh
    pip install -e /tmp/mkdocs-material
    ```

1. Copy `gen_docs.py` to `/tmp/mkdocs-material-bench-partials` and run it to generate a number of docs pages, e.g. 100, and `mkdocs.yml`:

    ```sh
    python gen_docs.py 100
    ```

1. Then, build the docs repeatedly, e.g. 50 times, and time it using, e.g., [`multitime`](https://tratt.net/laurie/src/multitime/):

    ```sh
    multitime -n 50 mkdocs build
    ```

1. Apply the patch:

    ```sh
    pushd /tmp/mkdocs-material
    git apply extract-partials.diff
    popd
    ```

    Then, repeat step 5 to measure the time again.

**Comment by squidfunk** [comment]
Wow, thanks for measuring! Always better to talk about numbers. My estimation is based on https://github.com/squidfunk/mkdocs-material/commit/81a13f6bc6bf0572fe106651988b4923bceea735, which introduced macros into `nav-item.html` because of bad performance on projects with many pages. The problem is, that rendering `nav-item.html` has quadratic complexity (every page links to every other page), while the header doesn't. Thus, my decision was based on an educated guess. Now that you crunched the numbers, we can reconsider.

Could you create a PR? The diff looks pretty good already, although it's a little hard to read.

**Comment by sisp** [comment]
Thanks for reconsidering! :pray: I'll create a PR. The diff will look the same there I guess, I extract the markup into a separate file and include it instead.

About https://github.com/squidfunk/mkdocs-material/commit/81a13f6bc6bf0572fe106651988b4923bceea735: I'm not so deeply familiar with Jinja's internal, but I wonder whether defining the macro in `nav-item.html` harms performance as it's included repeatedly, e.g.:

https://github.com/squidfunk/mkdocs-material/blob/a0921bcaf42251ecf92cd8a901c71030229b6c99/src/partials/nav.html#L60-L66

I wonder whether the performance would improve if the macro was defined outside `nav-item.html`. :thinking: Anyway, that's a different topic.

**Comment by squidfunk** [comment]
Thanks! I'll address the PR asap.

> I wonder whether the performance would improve if the macro was defined outside nav-item.html. 🤔 Anyway, that's a different topic.

Good question. If you manage to gather intel on this, feel free to share.

**Comment by squidfunk** [comment]
Released as part of [9.2.0](https://github.com/squidfunk/mkdocs-material/releases/tag/9.2.0). Thanks again for your work!
